[Docs] Creating feature vector over several feature sets with different entity (#2540)
mlrun/mlrun: docs/feature-store/feature-vectors.md

Defaults to the base feature set's timestamp entity.
- **target** &mdash; A Feature Store target to write the results to.  
Defaults to return as a return value to the caller.
- **run_config** &mdash; (optional) A function or a {py:class}`~mlrun.feature_store.RunConfig` to run the feature vector creation process in a remote function.
- **drop_columns** &mdash; (optional) A list of columns to drop from the resulting feature vector.
- **start_time** &mdash; (optional) Datetime, low limit of time needed to be filtered. 
- **end_time** &mdash; (optional) Datetime, high limit of time needed to be filtered. 
-**with_indexes**    return vector with index columns and timestamp_key from the feature sets. Default is False.
-**update_stats** &mdash; update features statistics from the requested feature sets on the vector. Default is False.
-**engine** &mdash; processing engine kind ("local", "dask", or "spark")
-**engine_args** &mdash; kwargs for the processing engine
-**query** &mdash; The query string used to filter rows
-**spark_service** &mdash; Name of the spark service to be used (when using a remote-spark runtime)   

> - **relations** &mdash; (optional) Dictionary that indicates all of the relations between different feature sets. It looks like: `{"feature_set_name_1:feature_set_name_2":{"column_of_1":"column_of_2",...}...}`. If the relation is None, and the `feature_set` 

Reviewer: remove this line - no "relations" in get_offline_features

> `get_offline_features` expects to receive:

Reviewer 2: I think you missing here few argument of this function : 

`:param with_indexes:    return vector with index columns and timestamp_key from the feature sets (default False)`
`:param update_stats:    update features statistics from the requested feature sets on the vector. Default is False.`
`:param engine:          processing engine kind ("local", "dask", or "spark")`
`:param engine_args:     kwargs for the processing engine`
`:param query:           The query string used to filter rows`
`:param spark_service:   Name of the spark service to be used (when using a remote-spark runtime)`

- **feature_vector** &mdash; A feature vector store reference or object.
- **entity_rows** &mdash; (optional) A dataframe that the features will be joined to. 
Defaults to the first feature set defined in the features vector's features list, and acts as the base for the vector's joins.
- **entity_timestamp_column** &mdash; (optional) A specific timestamp column (from the defined features) to act as the base timestamp column. 
Defaults to the base feature set's timestamp entity.
- **target** &mdash; A Feature Store target to write the results to.  
Defaults to return as a return value to the caller.
- **run_config** &mdash; (optional) A function or a {py:class}`~mlrun.feature_store.RunConfig` to run the feature vector creation process in a remote function.
- **drop_columns** &mdash; (optional) A list of columns to drop from the resulting feature vector.
- **start_time** &mdash; (optional) Datetime, low limit of time needed to be filtered. 
- **end_time** &mdash; (optional) Datetime, high limit of time needed to be filtered. 

> - **relations** &mdash; (optional) Dictionary that indicates all of the relations between different feature sets. It looks like: `{"feature_set_name_1:feature_set_name_2":{"column_of_1":"column_of_2",...}...}`. If the relation is None, and the `feature_set` 
> relations is also None, the join is done on the entity. Relevant only for Dask and storey(local) engines.<br>
> You can define relations of a feature set with the relations argument, like this:
> `{"feature_set_name": {"my_column":"other_feature_set_column", ...}...}`

Reviewer 2: ```suggestion

```

You can see a full example of using the offline feature vector to create an ML model in [part 2 of the end-to-end demo](./end-to-end-demo/02-create-training-model.html).

#### Using joins in an offline feature vector

You can create a join for:
- Feature sets that have a common entity
- Feature sets that do not have a common entity

**Feature sets that have a common entity**

In this case, the join is performed on the common entity.

```

> employees_set_entity = fs.Entity("id")
>         employees_set = fs.FeatureSet(
>             "employees",
>             entities=[employees_set_entity],
>             relations={"department_id": departments_set_entity},
>         )
>         employees_set.set_targets(targets=["parquet"], with_defaults=False)
>         fs.ingest(employees_set, employees)
> 
>         mini_employees_set = fs.FeatureSet(
>             "mini-employees",
>             entities=[employees_set_entity],
>             relations={
>                 "department_id": departments_set_entity,
>                 "class_id": classes_set_entity,
>             },
>         )
>         mini_employees_set.set_targets(targets=["parquet"], with_defaults=False)
>         fs.ingest(mini_employees_set, employees_mini)
> 
>         features = ["employees.name as n", "mini-employees.name as mini_name"]
> 
>         vector_3 = fs.FeatureVector(
>             "mini-emp-vec", features, description="Employees feature vector"
>         )
>         vector_3.save()
> 
>         resp_3 = fs.get_offline_features(
>             vector_3,
>             join_type=join_type,
>             engine_args=engine_args,
>             with_indexes=with_indexes,
>             engine=engine,
>         )

Reviewer 2: ```suggestion
employees_set_entity = fs.Entity("id")
employees_set = fs.FeatureSet(
    "employees",
    entities=[employees_set_entity],
    relations={"department_id": departments_set_entity},
)
employees_set.set_targets(targets=["parquet"], with_defaults=False)
fs.ingest(employees_set, employees)

mini_employees_set = fs.FeatureSet(
    "mini-employees",
    entities=[employees_set_entity],
    relations={
        "department_id": departments_set_entity,
        "class_id": classes_set_entity,
    },
)
mini_employees_set.set_targets(targets=["parquet"], with_defaults=False)
fs.ingest(mini_employees_set, employees_mini)

features = ["employees.name as n", "mini-employees.name as mini_name"]

vector = fs.FeatureVector(
    "mini-emp-vec", features, description="Employees feature vector"
)
vector.save()

resp = fs.get_offline_features(
    vector,
    join_type='outer', # one of following values: "inner" (as with current code), "outer", "right", "left"
    engine_args=engine_args,
    with_indexes=True,
)
```

```

**Feature sets that do not have a common entity**

In this case, you define the relations between the features set with the argument: ` relations={column_name(str): Entity}`</br>
and you include this dictionary when initializing the feature set. 

```

> departments_set_entity = fs.Entity("d_id")
> departments_set = fs.FeatureSet(
>     "departments",
>     entities=[departments_set_entity],
> )
> 
> departments_set.set_targets(targets=["parquet"], with_defaults=False)
> fs.ingest(departments_set, departments)
> 
> employees_set_entity = fs.Entity("id")
> employees_set = fs.FeatureSet(
>     "employees",
>     entities=[employees_set_entity],
>     relations={"department_id": departments_set_entity},  # dictionary where the key is str identifying a column/feature on this feature-set, and the dictionary value is an Entity object on another feature-set
> )
> employees_set.set_targets(targets=["parquet"], with_defaults=False)
> fs.ingest(employees_set, employees)
> features = ["employees.name as emp_name", "departments.name as dep_name"]
> 
> vector = fs.FeatureVector(
>     "employees-vec", features, description="Employees feature vector"
> )
> 
> resp_1 = fs.get_offline_features(
>     vector,
>     join_type='inner', # one of following values: "inner" (as with current code), "outer", "right", "left"
>     engine_args=engine_args,
>     with_indexes=with_indexes,
>     engine=engine,
> )

Reviewer 2: ```suggestion
departments_set_entity = fs.Entity("d_id")
departments_set = fs.FeatureSet(
    "departments",
    entities=[departments_set_entity],
)

departments_set.set_targets(targets=["parquet"], with_defaults=False)
fs.ingest(departments_set, departments)

employees_set_entity = fs.Entity("id")
employees_set = fs.FeatureSet(
    "employees",
    entities=[employees_set_entity],
    relations={"department_id": departments_set_entity},  # dictionary where the key is str identifying a column/feature on this feature-set, and the dictionary value is an Entity object on another feature-set
)
employees_set.set_targets(targets=["parquet"], with_defaults=False)
fs.ingest(employees_set, employees)
features = ["employees.name as emp_name", "departments.name as dep_name"]

vector = fs.FeatureVector(
    "employees-vec", features, description="Employees feature vector"
)

resp = fs.get_offline_features(
    vector,
    join_type='inner', # one of following values: "inner" (as with current code), "outer", "right", "left"
    engine_args=engine_args,
    with_indexes=False,
)
```

You can see a full example of using the offline feature vector to create an ML model in [part 2 of the end-to-end demo](./end-to-end-demo/02-create-training-model.html).

#### Using joins in an offline feature vector

You can create a join for:
- Feature sets that have a common entity
- Feature sets that do not have a common entity

**Feature sets that have a common entity**

In this case, the join is performed on the common entity.

```
employees_set_entity = fs.Entity("id")
employees_set = fs.FeatureSet(
    "employees",
    entities=[employees_set_entity],

>     relations={"department_id": departments_set_entity},

Reviewer: the join in this example is entity, remove the relations dictionary

)
employees_set.set_targets(targets=["parquet"], with_defaults=False)
fs.ingest(employees_set, employees)

mini_employees_set = fs.FeatureSet(
    "mini-employees",
    entities=[employees_set_entity],

>     relations={

Reviewer: the join in this example is entity, remove the relations dictionary
